### PR TITLE
Cherry-pick annotation based TD selection to `release`

### PR DESCRIPTION
**Pull request by blindpirate**
Cherry-picks:

- https://github.com/gradle/gradle/pull/22431
- https://github.com/gradle/gradle/pull/22225
- https://github.com/gradle/gradle/pull/22223

### CODE REVIEW

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

File: .teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt [hunk 1]
```diff
@@ -202,16 +202,14 @@ class FunctionalTestBucketGenerator(private val model: CIBuildModel, testTimeDat
...
             testCoverage.testType == TestType.platform ->
                 splitDocsSubproject(validSubprojects) +
                     splitIntoBuckets(validSubprojects, subProjectTestClassTimes, testCoverage, listOf("docs"), false)
             testCoverage.os == Os.LINUX ->
-                splitIntoBuckets(validSubprojects, subProjectTestClassTimes, testCoverage, listOf("file-watching"), true) +
-                    SmallSubprojectBucket(validSubprojects.first { it.name == "file-watching" }, false)
+                splitIntoBuckets(validSubprojects, subProjectTestClassTimes, testCoverage, emptyList(), true)
             else ->
-                splitIntoBuckets(validSubprojects, subProjectTestClassTimes, testCoverage, listOf("file-watching"), false)
+                splitIntoBuckets(validSubprojects, subProjectTestClassTimes, testCoverage, emptyList(), false)
```
Comment (wolfs) [inline .teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt:212]: I was wondering if it is worth picking the changes to the TC buckets as well. For example I am not sure which test data is better for `release`, the current one or the one we now have on `master`. We could for example only cherry-pick the support for using the TD annotations without touching the buckets.

I don't have a strong opinion on the above, only wanted to note that this is an option as well, so we change fewer things on `release`. Or maybe it is better to pick as much as possible, given that `release` will become the LTS branch.

**Review by wolfs — APPROVED** [review]
LGTM!

**Pull request merged**